src/xlsx/xlsx_writer.ts

```diff
@@ -361,3 +362,40 @@ function createRelRoot(): XLSXExportFile {
   `;
   return createXMLFile(parseXML(xml), "_rels/.rels");
 }
+
+/**
+ * Excel sheet names are maximum 31 characters while o-spreadsheet do not have this limit.
+ * This method converts the sheet names to be within the 31 characters limit.
+ * The cells/charts referencing this sheet will be updated accordingly.
+ */
+export function fixLengthySheetNames(data: ExcelWorkbookData): ExcelWorkbookData {
+  const nameMapping: Record<string, string> = {};
+  const newNames = new Set<string>();
+  for (const sheet of data.sheets) {
+    let newName = sheet.name.slice(0, 31);
+    let i = 1;
+    while (newNames.has(newName)) {
+      newName = newName.slice(0, 31 - String(i).length) + i++;
+    }
+    newNames.add(newName);
+    if (newName !== sheet.name) {
+      nameMapping[sheet.name] = newName;
+      sheet.name = newName;
+    }
+  }
+
+  if (!newNames.size) {
+    return data;
+  }
+
+  const sheetWithNewNames = Object.keys(nameMapping).sort((a, b) => b.length - a.length);
+  let stringifiedData = JSON.stringify(data);
+  for (const sheetName of sheetWithNewNames) {
+    const regex = new RegExp(`'?${sheetName}'?!`, "g");
+    stringifiedData = stringifiedData.replaceAll(regex, (match) => {
+      const newName = nameMapping[sheetName];
+      return match.replace(sheetName, newName);
+    });
```

Comment:
I wanted the additional security of only matching sheet name + (maybe quotation marks) + exclamation point instead of just the sheet name. And I need a regex for that, because the quotation mark are optional :x